src/librbd/deep_copy/ObjectCopyRequest.cc

```diff
@@ -0,0 +1,632 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "ObjectCopyRequest.h"
+#include "common/errno.h"
+#include "librados/snap_set_diff.h"
+#include "librbd/ExclusiveLock.h"
+#include "librbd/ObjectMap.h"
+#include "librbd/Utils.h"
+#include "osdc/Striper.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::deep_copy::ObjectCopyRequest: " \
+                           << this << " " << __func__ << ": "
+
+namespace librados {
+
+inline bool operator==(const clone_info_t& rhs, const clone_info_t& lhs) {
+  return (rhs.cloneid == lhs.cloneid &&
+          rhs.snaps == lhs.snaps &&
+          rhs.overlap == lhs.overlap &&
+          rhs.size == lhs.size);
+}
+
+inline bool operator==(const snap_set_t& rhs, const snap_set_t& lhs) {
+  return (rhs.clones == lhs.clones &&
+          rhs.seq == lhs.seq);
+}
+
+} // namespace librados
+
+namespace librbd {
+namespace deep_copy {
+
+using librbd::util::create_context_callback;
+using librbd::util::create_rados_callback;
+
+template <typename I>
+ObjectCopyRequest<I>::ObjectCopyRequest(I *src_image_ctx, I *dst_image_ctx,
+                                        const SnapMap &snap_map,
+                                        uint64_t dst_object_number,
+                                        Context *on_finish)
+  : m_src_image_ctx(src_image_ctx), m_dst_image_ctx(dst_image_ctx),
+    m_cct(dst_image_ctx->cct), m_snap_map(snap_map),
+    m_dst_object_number(dst_object_number), m_on_finish(on_finish) {
+  assert(!snap_map.empty());
+
+  m_src_io_ctx.dup(m_src_image_ctx->data_ctx);
+  m_dst_io_ctx.dup(m_dst_image_ctx->data_ctx);
+
+  m_dst_oid = m_dst_image_ctx->get_object_name(dst_object_number);
+
+  ldout(m_cct, 20) << "dst_oid=" << m_dst_oid << dendl;
+}
+
+template <typename I>
+void ObjectCopyRequest<I>::send() {
+  init_offsets();
+  send_list_snaps();
+}
+
+template <typename I>
+void ObjectCopyRequest<I>::send_list_snaps() {
+  // starting copy from highest src object number is important for truncate
```

Comment:
If a destination object number `x` is mapped to src object numbers `a, b, c` (due to different object sizes) I assume that it will be always `a < b < c` for any striping. Now, if reading src objects I see that `c` does not exist, and `b` is truncated, I can truncate the destination object up to `b` truncated length. If `b` is truncated but `c` exists I know that I need to write a hole. 

For this reason I start listing src objects from the highest one.

Reply:
If fancy striping is used, I am not sure that logic is correct:

src:
- object size 2
- stripe unit = 1
- stripe count = 3

dst:
- object size 4
- no fancy striping

The mapping would be: object 1@0~1, object 2@0~1, object 3@0~1, object 1@1~1. If object 3 doesn't exist in this case, that's a hole and not a truncate.

Reply:
@dillaman It is good to know. Thanks!